### ISSUE TITLE
Keyboard hangs/freezes for couple of seconds

### ISSUE THREAD

**Issue by Svalorzen** [tmk/tmk_keyboard#541]
Hi, I've recently set up the tmk_keyboard using this guide: https://deskthority.net/workshop-f7/how-to-build-your-very-own-keyboard-firmware-t7177.html

From time to time unfortunately it seems that the keyboard hangs for a couple of seconds. This results in either keys not working, or in the same key being repeated a couple hundred times.

The keyboard used to have another chip before I switched to a Teensy++ 2.0 and tmk, and this particular problem was never present.

Aside from that, the keyboard works perfectly and all keys are recognized. Debugging also works, however I'm not sure how I would start to debug this problem as it is random and not easy to reproduce.

Any suggestions on what could be causing this, and how to approach it?

**Comment by Svalorzen** [comment]
Just in case, I'm adding here the diff with my changes, if you feel like checking it out:

    diff --git a/keyboard/gh60/Makefile b/keyboard/gh60/Makefile
    index b0e7a30e..01e9d97f 100644
    --- a/keyboard/gh60/Makefile
    +++ b/keyboard/gh60/Makefile
    @@ -27,7 +27,7 @@
     # make flip-ee = Download the eeprom file to the device, using Atmel FLIP
     #                (must have Atmel FLIP installed).
     #
    -# make debug = Start either simulavr or avarice as specified for debugging, 
    +# make debug = Start either simulavr or avarice as specified for debugging,
     #              with avr-gdb or avr-insight as the front end for debugging.
     #
     # make filename.s = Just compile filename.c into the assembler code only.
    @@ -61,8 +61,8 @@ CONFIG_H = config.h
     
     
     # MCU name
    -#MCU = at90usb1287
    -MCU = atmega32u4
    +MCU = at90usb1286
    +#MCU = atmega32u4
     
     # Processor frequency.
     #     This will define a symbol, F_CPU, in all source code files equal to the
    @@ -119,7 +119,7 @@ EXTRAKEY_ENABLE = yes	# Audio control and System control(+450)
     CONSOLE_ENABLE = yes	# Console for debug(+400)
     COMMAND_ENABLE = yes    # Commands for debug and configuration
     #SLEEP_LED_ENABLE = yes  # Breathing sleep LED during USB suspend
    -NKRO_ENABLE = yes	# USB Nkey Rollover
    +# NKRO_ENABLE = yes	# USB Nkey Rollover
     
     
     # Optimize size but this may cause error "relocation truncated to fit"
    diff --git a/keyboard/gh60/config.h b/keyboard/gh60/config.h
    index e9c0f436..b5788ab1 100644
    --- a/keyboard/gh60/config.h
    +++ b/keyboard/gh60/config.h
    @@ -28,8 +28,8 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
     #define DESCRIPTION     t.m.k. keyboard firmware for GH60
     
     /* key matrix size */
    -#define MATRIX_ROWS 5
    -#define MATRIX_COLS 14
    +#define MATRIX_ROWS 12
    +#define MATRIX_COLS 11
     
     /* define if matrix has ghost */
     //#define MATRIX_HAS_GHOST
    diff --git a/keyboard/gh60/keymap_common.h b/keyboard/gh60/keymap_common.h
    index 9ac57221..3ec596b5 100644
    --- a/keyboard/gh60/keymap_common.h
    +++ b/keyboard/gh60/keymap_common.h
    @@ -33,17 +33,33 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
      * K2C, K31 and  K3C are extra keys for ISO
      */
     #define KEYMAP( \
    -    K00, K01, K02, K03, K04, K05, K06, K07, K08, K09, K0A, K0B, K0C, K0D, \
    -    K10, K11, K12, K13, K14, K15, K16, K17, K18, K19, K1A, K1B, K1C, K1D, \
    -    K20, K21, K22, K23, K24, K25, K26, K27, K28, K29, K2A, K2B, K2C, K2D, \
    -    K30, K31, K32, K33, K34, K35, K36, K37, K38, K39, K3A, K3B, K3C, K3D, \
    -    K40, K41, K42,           K45,                K49, K4A, K4B, K4C, K4D  \
    +    K00,      K02, K03, K04, K05, K06, K07, K08, K09,      \
    +    K10, K11, K12, K13, K14, K15, K16, K17, K18, K19, K1A, \
    +    K20,      K22, K23, K24, K25, K26, K27, K28, K29, K2A, \
    +    K30,      K32, K33, K34, K35, K36, K37, K38, K39, K3A, \
    +    K40,      K42, K43, K44, K45, K46, K47, K48, K49, K4A, \
    +    K50, K51, K52,                K56,                K5A, \
    +                                                           \
    +    K60, K61, K62, K63, K64, K65, K66,                     \
    +    K70, K71,      K73, K74, K75, K76, K77, K78, K79, K7A, \
    +    K80, K81, K82,      K84, K85, K86, K87, K88, K89,      \
    +    K90, K91, K92, K93,      K95,      K97, K98, K99, K9A, \
    +    KA0, KA1,      KA3,      KA5,      KA7, KA8, KA9,      \
    +    KB0,      KB2, KB3, KB4, KB5, KB6, KB7,      KB9, KBA  \
     ) { \
    -    { KC_##K00, KC_##K01, KC_##K02, KC_##K03, KC_##K04, KC_##K05, KC_##K06, KC_##K07, KC_##K08, KC_##K09, KC_##K0A, KC_##K0B, KC_##K0C, KC_##K0D }, \
    -    { KC_##K10, KC_##K11, KC_##K12, KC_##K13, KC_##K14, KC_##K15, KC_##K16, KC_##K17, KC_##K18, KC_##K19, KC_##K1A, KC_##K1B, KC_##K1C, KC_##K1D }, \
    -    { KC_##K20, KC_##K21, KC_##K22, KC_##K23, KC_##K24, KC_##K25, KC_##K26, KC_##K27, KC_##K28, KC_##K29, KC_##K2A, KC_##K2B, KC_##K2C, KC_##K2D }, \
    -    { KC_##K30, KC_##K31, KC_##K32, KC_##K33, KC_##K34, KC_##K35, KC_##K36, KC_##K37, KC_##K38, KC_##K39, KC_##K3A, KC_##K3B, KC_##K3C, KC_##K3D }, \
    -    { KC_##K40, KC_##K41, KC_##K42, KC_NO,    KC_NO,    KC_##K45, KC_NO,    KC_NO,    KC_NO,    KC_##K49, KC_##K4A, KC_##K4B, KC_##K4C, KC_##K4D }  \
    +    { KC_##K00, KC_NO   , KC_##K02, KC_##K03, KC_##K04, KC_##K05, KC_##K06, KC_##K07, KC_##K08, KC_##K09, KC_NO    }, \
    +    { KC_##K10, KC_##K11, KC_##K12, KC_##K13, KC_##K14, KC_##K15, KC_##K16, KC_##K17, KC_##K18, KC_##K19, KC_##K1A }, \
    +    { KC_##K20, KC_NO   , KC_##K22, KC_##K23, KC_##K24, KC_##K25, KC_##K26, KC_##K27, KC_##K28, KC_##K29, KC_##K2A }, \
    +    { KC_##K30, KC_NO   , KC_##K32, KC_##K33, KC_##K34, KC_##K35, KC_##K36, KC_##K37, KC_##K38, KC_##K39, KC_##K3A }, \
    +    { KC_##K40, KC_NO   , KC_##K42, KC_##K43, KC_##K44, KC_##K45, KC_##K46, KC_##K47, KC_##K48, KC_##K49, KC_##K4A }, \
    +    { KC_##K50, KC_##K51, KC_##K52, KC_NO   , KC_NO   , KC_NO   , KC_##K56, KC_NO   , KC_NO,    KC_NO   , KC_##K5A }, \
    +                                                                                                                      \
    +    { KC_##K60, KC_##K61, KC_##K62, KC_##K63, KC_##K64, KC_##K65, KC_##K66, KC_NO   , KC_NO   , KC_NO   , KC_NO    }, \
    +    { KC_##K70, KC_##K71, KC_NO   , KC_##K73, KC_##K74, KC_##K75, KC_##K76, KC_##K77, KC_##K78, KC_##K79, KC_##K7A }, \
    +    { KC_##K80, KC_##K81, KC_##K82, KC_NO   , KC_##K84, KC_##K85, KC_##K86, KC_##K87, KC_##K88, KC_##K89, KC_NO    }, \
    +    { KC_##K90, KC_##K91, KC_##K92, KC_##K93, KC_NO   , KC_##K95, KC_NO   , KC_##K97, KC_##K98, KC_##K99, KC_##K9A }, \
    +    { KC_##KA0, KC_##KA1, KC_NO   , KC_##KA3, KC_NO   , KC_##KA5, KC_NO   , KC_##KA7, KC_##KA8, KC_##KA9, KC_NO    }, \
    +    { KC_##KB0, KC_NO   , KC_##KB2, KC_##KB3, KC_##KB4, KC_##KB5, KC_##KB6, KC_##KB7, KC_NO,    KC_##KB9, KC_##KBA }  \
     }
     
     /* ANSI variant. No extra keys for ISO */
    diff --git a/keyboard/gh60/keymap_poker.c b/keyboard/gh60/keymap_poker.c
    index 5e94184f..efd71c7e 100644
    --- a/keyboard/gh60/keymap_poker.c
    +++ b/keyboard/gh60/keymap_poker.c
    @@ -2,103 +2,47 @@
     
     const uint8_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
         /* 0: qwerty */
    -    KEYMAP_ANSI(
    -        GRV, 1,   2,   3,   4,   5,   6,   7,   8,   9,   0,   MINS,EQL, BSPC, \
    -        TAB, Q,   W,   E,   R,   T,   Y,   U,   I,   O,   P,   LBRC,RBRC,BSLS, \
    -        CAPS,A,   S,   D,   F,   G,   H,   J,   K,   L,   SCLN,QUOT,     ENT,  \
    -        LSFT,Z,   X,   C,   V,   B,   N,   M,   COMM,DOT, SLSH,          RSFT, \
    -        LCTL,LGUI,LALT,          SPC,                     FN0, RGUI,APP, RCTL),
    -    /* 1: colemak */
    -    KEYMAP_ANSI(
    -        GRV, 1,   2,   3,   4,   5,   6,   7,   8,   9,   0,   MINS,EQL, BSPC, \
    -        TAB, Q,   W,   F,   P,   G,   J,   L,   U,   Y,   SCLN,LBRC,RBRC,BSLS, \
    -        BSPC,A,   R,   S,   T,   D,   H,   N,   E,   I,   O,   QUOT,     ENT,  \
    -        LSFT,Z,   X,   C,   V,   B,   K,   M,   COMM,DOT, SLSH,          RSFT, \
    -        LCTL,LGUI,LALT,          SPC,                     FN0, RGUI,APP, RCTL),
    -    /* 2: dvorak */
    -    KEYMAP_ANSI(
    -        GRV, 1,   2,   3,   4,   5,   6,   7,   8,   9,   0,   LBRC,RBRC,BSPC, \
    -        TAB, QUOT,COMM,DOT, P,   Y,   F,   G,   C,   R,   L,   SLSH,EQL, BSLS, \
    -        CAPS,A,   O,   E,   U,   I,   D,   H,   T,   N,   S,   MINS,     ENT,  \
    -        LSFT,SCLN,Q,   J,   K,   X,   B,   M,   W,   V,   Z,             RSFT, \
    -        LCTL,LGUI,LALT,          SPC,                     FN0, RGUI,APP, RCTL),
    -    /* 3: workman */
    -    KEYMAP_ANSI(
    -        GRV, 1,   2,   3,   4,   5,   6,   7,   8,   9,   0,   MINS,EQL, BSPC, \
    -        TAB, Q,   D,   R,   W,   B,   J,   F,   U,   P,   SCLN,LBRC,RBRC,BSLS, \
    -        BSPC,A,   S,   H,   T,   G,   Y,   N,   E,   O,   I,   QUOT,     ENT,  \
    -        LSFT,Z,   X,   M,   C,   V,   K,   L,   COMM,DOT, SLSH,          RSFT, \
    -        LCTL,LGUI,LALT,          SPC,                     FN0, RGUI,APP, RCTL),
    -    /* 4: Poker with Arrow */
    -    KEYMAP_ANSI(
    -        TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS, \
    -        TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS, \
    -        TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,     TRNS, \
    -        TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,          UP,   \
    -        TRNS,TRNS,TRNS,          TRNS,                    TRNS,LEFT,DOWN,RGHT),
    -    /* 5: Poker with Esc */
    -    KEYMAP_ANSI(
    -        ESC, TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS, \
    -        TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS, \
    -        TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,     TRNS, \
    -        TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,          TRNS, \
    -        TRNS,TRNS,TRNS,          TRNS,                    TRNS,TRNS,TRNS,TRNS),
    -    /* 6: Poker Fn
    -     * ,-----------------------------------------------------------.
    -     * |Esc| F1| F2| F3| F4| F5| F6| F7| F8| F9|F10|F11|F12|       |
    -     * |-----------------------------------------------------------|
    -     * |     |FnQ| Up|   |   |   |   |   |   |Cal|   |Hom|Ins|FnL  |
    -     * |-----------------------------------------------------------|
    -     * |      |Lef|Dow|Rig|   |   |Psc|Slk|Pau|   |Tsk|End|        |
    -     * |-----------------------------------------------------------|
    -     * |        |Del|   |Web|Mut|VoU|VoD|   |PgU|PgD|Del|          |
    -     * |-----------------------------------------------------------|
    -     * |    |    |    |         FnS            |    |    |    |    |
    -     * `-----------------------------------------------------------'
    -     * Fn:  to Fn overlay
    -     * FnL: to Layout selector overaly
    -     * FnQ: toggle Esc overlay
    -     * FnS: toggle Arrow overlay
    -     */
    -    KEYMAP_ANSI(
    -        ESC, F1,  F2,  F3,  F4,  F5,  F6,  F7,  F8,  F9,  F10, F11, F12, TRNS, \
    -        TRNS,FN2, UP,  TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,CALC,TRNS,HOME,INS, FN4,  \
    -        TRNS,LEFT,DOWN,RGHT,TRNS,TRNS,PSCR,SLCK,PAUS,TRNS,FN3, END,      TRNS, \
    -        TRNS,DEL, TRNS,WHOM,MUTE,VOLU,VOLD,TRNS,PGUP,PGDN,DEL,           TRNS, \
    -        TRNS,TRNS,TRNS,          FN1,                     TRNS,TRNS,TRNS,TRNS),
    -    /* 7: Layout selector
    -     * ,-----------------------------------------------------------.
    -     * | Lq| Lc| Ld| Lw|   |   |   |   |   |   |   |   |   |       |
    -     * |-----------------------------------------------------------|
    -     * |     |Lq |Lw |   |   |   |   |   |   |   |   |   |   |     |
    -     * |-----------------------------------------------------------|
    -     * |      |   |   |Ld |   |   |   |   |   |   |   |   |        |
    -     * |-----------------------------------------------------------|
    -     * |        |   |   |Lc |   |   |   |   |   |   |   |          |
    -     * |-----------------------------------------------------------|
    -     * |    |    |    |                        |    |    |    |    |
    -     * `-----------------------------------------------------------'
    -     * Lq: set Qwerty layout
    -     * Lc: set Colemak layout
    -     * Ld: set Dvorak layout
    -     * Lw: set Workman layout
    +    KEYMAP(
    +         ESC,         F1,   F2,   F3,   F4,   F5,   F6,   F7,   F8,       \
    +         GRV,    1,    2,    3,    4,    5,    6,    7,    8,    9,    0, \
    +         TAB,          Q,    W,    E,    R,    T,    Y,    U,    I,    O, \
    +        CAPS,          A,    S,    D,    F,    G,    H,    J,    K,    L, \
    +        LSFT,          Z,    X,    C,    V,    B,    N,    M, COMM,  DOT, \
    +        LCTL, LGUI, LALT,                    SPC,                   RALT, \
    +                                                                          \
    +          F9,  F10,  F11,  F12, SLCK, PSCR, PAUS,                         \
    +        MINS,  EQL,       BSPC,  INS, HOME, PGUP, NLCK, PSLS, PAST, PMNS, \
    +           P, LBRC, RBRC,        DEL,  END, PGDN,   P7,   P8,   P9,       \
    +        SCLN, QUOT,    T,  ENT,         UP,         P4,   P5,   P6, PPLS, \
    +        SLSH, BSLS,       RSFT,         UP,         P1,   P2,   P3,       \
    +        RGUI,       MENU, RCTL, LEFT, DOWN, RGHT,   P0,       PDOT, PENT
    +    ),
    +
    +    /* 0: qwerty
    +     * ,----------------------------------------------------------------------------------------.
    +     * |Esc|   | F1| F2| F3| F4| F5| F6| F7| F8|   || F9|F10|F11|F12|ScL|Prt|Pau|   |   |   |   |
    +     * |----------------------------------------------------------------------------------------|
    +     * | ` | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 0 || - | = |   |BkS|Ins|Hom|PUp|NmL| / | * | - |
    +     * |----------------------------------------------------------------------------------------|
    +     * |Tab|   | Q | W | E | R | T | Y | U | I | O || P | [ | ] |   |Del|End|PDw| 7 | 8 | 9 |   |
    +     * |----------------------------------------------------------------------------------------|
    +     * |CpL|   | A | S | D | F | G | H | J | K | L || ; | ' | & |Ent|   |MUp|   | 4 | 5 | 6 | + |
    +     * |----------------------------------------------------------------------------------------|
    +     * |Shf|   | Z | X | C | V | B | N | M | < | > || / | \ |   |Shf|   | Up|   | 1 | 2 | 3 |   |
    +     * |----------------------------------------------------------------------------------------|
    +     * |Ctr|Sup|Alt|   |   |   |Spc|   |   |   |Alt||Sup|   |Hyp|Ctr|Lft|Dwn|Rht| 0 |   | . |Ent|
    +     * `----------------------------------------------------------------------------------------'
          */
    -    KEYMAP_ANSI(
    -        FN5, FN6, FN7, FN8, TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS, \
    -        TRNS,FN5, FN8, TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS, \
    -        TRNS,TRNS,TRNS,FN7, TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,     TRNS, \
    -        TRNS,TRNS,TRNS,FN6, TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,TRNS,          TRNS, \
    -        TRNS,TRNS,TRNS,          TRNS,                    TRNS,TRNS,TRNS,TRNS),
     };
     const action_t PROGMEM fn_actions[] = {
         /* Poker Layout */
    -    [0] = ACTION_LAYER_MOMENTARY(6),  // to Fn overlay
    -    [1] = ACTION_LAYER_TOGGLE(4),     // toggle arrow overlay
    -    [2] = ACTION_LAYER_TOGGLE(5),     // toggle Esc overlay
    -    [3] = ACTION_MODS_KEY(MOD_RCTL|MOD_RSFT, KC_ESC), // Task(RControl,RShift+Esc)
    -    [4] = ACTION_LAYER_MOMENTARY(7),  // to Layout selector
    -    [5] = ACTION_DEFAULT_LAYER_SET(0),  // set qwerty layout
    -    [6] = ACTION_DEFAULT_LAYER_SET(1),  // set colemak layout
    -    [7] = ACTION_DEFAULT_LAYER_SET(2),  // set dvorak layout
    -    [8] = ACTION_DEFAULT_LAYER_SET(3),  // set workman layout
    +    //[0] = ACTION_LAYER_MOMENTARY(6),  // to Fn overlay
    +    //[1] = ACTION_LAYER_TOGGLE(4),     // toggle arrow overlay
    +    //[2] = ACTION_LAYER_TOGGLE(5),     // toggle Esc overlay
    +    //[3] = ACTION_MODS_KEY(MOD_RCTL|MOD_RSFT, KC_ESC), // Task(RControl,RShift+Esc)
    +    //[4] = ACTION_LAYER_MOMENTARY(7),  // to Layout selector
    +    //[5] = ACTION_DEFAULT_LAYER_SET(0),  // set qwerty layout
    +    //[6] = ACTION_DEFAULT_LAYER_SET(1),  // set colemak layout
    +    //[7] = ACTION_DEFAULT_LAYER_SET(2),  // set dvorak layout
    +    //[8] = ACTION_DEFAULT_LAYER_SET(3),  // set workman layout
     };
    diff --git a/keyboard/gh60/led.c b/keyboard/gh60/led.c
    index 62abcd79..b781c53d 100644
    --- a/keyboard/gh60/led.c
    +++ b/keyboard/gh60/led.c
    @@ -22,13 +22,33 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
     
     void led_set(uint8_t usb_led)
     {
    -    if (usb_led & (1<<USB_LED_CAPS_LOCK)) {
    -        // output low
    -        DDRB |= (1<<2);
    -        PORTB &= ~(1<<2);
    -    } else {
    -        // Hi-Z
    -        DDRB &= ~(1<<2);
    -        PORTB &= ~(1<<2);
    -    }
    +    //if (usb_led & (1<<USB_LED_CAPS_LOCK)) {
    +    //    // output low
    +    //    DDRB |= (1<<2);
    +    //    PORTB &= ~(1<<2);
    +    //} else {
    +    //    // Hi-Z
    +    //    DDRB &= ~(1<<2);
    +    //    PORTB &= ~(1<<2);
    +    //}
    +
    +    //if (usb_led & (1<<USB_LED_SCROLL_LOCK)) {
    +    //    // output low
    +    //    DDRB |= (1<<2);
    +    //    PORTB &= ~(1<<2);
    +    //} else {
    +    //    // Hi-Z
    +    //    DDRB &= ~(1<<2);
    +    //    PORTB &= ~(1<<2);
    +    //}
    +
    +    //if (usb_led & (1<<USB_LED_NUM_LOCK)) {
    +    //    // output low
    +    //    DDRB |= (1<<2);
    +    //    PORTB &= ~(1<<2);
    +    //} else {
    +    //    // Hi-Z
    +    //    DDRB &= ~(1<<2);
    +    //    PORTB &= ~(1<<2);
    +    //}
     }
    diff --git a/keyboard/gh60/matrix.c b/keyboard/gh60/matrix.c
    index 5a56cee0..31db9cd6 100644
    --- a/keyboard/gh60/matrix.c
    +++ b/keyboard/gh60/matrix.c
    @@ -99,34 +99,31 @@ matrix_row_t matrix_get_row(uint8_t row)
     static void  init_cols(void)
     {
         // Input with pull-up(DDR:0, PORT:1)
    -    DDRF  &= ~(1<<0 | 1<<1);
    -    PORTF |=  (1<<0 | 1<<1);
    -    DDRE  &= ~(1<<6);
    -    PORTE |=  (1<<6);
    -    DDRD  &= ~(1<<7 | 1<<6 | 1<<4);
    -    PORTD |=  (1<<7 | 1<<6 | 1<<4);
    -    DDRC  &= ~(1<<7 | 1<<6);
    -    PORTC |=  (1<<7 | 1<<6);
    -    DDRB  &= ~(1<<7 | 1<<6 | 1<< 5 | 1<<4 | 1<<3 | 1<<1 | 1<<0);
    -    PORTB |=  (1<<7 | 1<<6 | 1<< 5 | 1<<4 | 1<<3 | 1<<1 | 1<<0);
    +    //DDRF  &= ~(1<<0 | 1<<1);
    +    //PORTF |=  (1<<0 | 1<<1);
    +    //DDRE  &= ~(1<<6);
    +    //PORTE |=  (1<<6);
    +    DDRC  &= ~(1<<6 | 1<<5 | 1<<4 | 1<<3 | 1<<2 | 1<<1 | 1<<0);
    +    PORTC |=  (1<<6 | 1<<5 | 1<<4 | 1<<3 | 1<<2 | 1<<1 | 1<<0);
    +    DDRD  &= ~(1<<3 | 1<<2 | 1<<1 | 1<<0);
    +    PORTD |=  (1<<3 | 1<<2 | 1<<1 | 1<<0);
    +    //DDRB  &= ~(1<<7 | 1<<6 | 1<< 5 | 1<<4 | 1<<3 | 1<<1 | 1<<0);
    +    //PORTB |=  (1<<7 | 1<<6 | 1<< 5 | 1<<4 | 1<<3 | 1<<1 | 1<<0);
     }
     
     static matrix_row_t read_cols(void)
     {
    -    return (PINF&(1<<0) ? 0 : (1<<0)) |
    -           (PINF&(1<<1) ? 0 : (1<<1)) |
    -           (PINE&(1<<6) ? 0 : (1<<2)) |
    -           (PINC&(1<<7) ? 0 : (1<<3)) |
    -           (PINC&(1<<6) ? 0 : (1<<4)) |
    -           (PINB&(1<<6) ? 0 : (1<<5)) |
    -           (PIND&(1<<4) ? 0 : (1<<6)) |
    -           (PINB&(1<<1) ? 0 : (1<<7)) |
    -           ((PINB&(1<<0) && PINB&(1<<7)) ? 0 : (1<<8)) |     // Rev.A and B
    -           (PINB&(1<<5) ? 0 : (1<<9)) |
    -           (PINB&(1<<4) ? 0 : (1<<10)) |
    -           (PIND&(1<<7) ? 0 : (1<<11)) |
    -           (PIND&(1<<6) ? 0 : (1<<12)) |
    -           (PINB&(1<<3) ? 0 : (1<<13));
    +    return (PINC&(1<<6) ? 0 : (1<<0)) |
    +           (PINC&(1<<5) ? 0 : (1<<1)) |
    +           (PINC&(1<<4) ? 0 : (1<<2)) |
    +           (PINC&(1<<3) ? 0 : (1<<3)) |
    +           (PINC&(1<<2) ? 0 : (1<<4)) |
    +           (PINC&(1<<1) ? 0 : (1<<5)) |
    +           (PINC&(1<<0) ? 0 : (1<<6)) |
    +           (PIND&(1<<3) ? 0 : (1<<7)) |
    +           (PIND&(1<<2) ? 0 : (1<<8)) |
    +           (PIND&(1<<1) ? 0 : (1<<9)) |
    +           (PIND&(1<<0) ? 0 : (1<<10));
     }
     
     /* Row pin configuration
    @@ -136,8 +133,10 @@ static matrix_row_t read_cols(void)
     static void unselect_rows(void)
     {
         // Hi-Z(DDR:0, PORT:0) to unselect
    -    DDRD  &= ~0b00101111;
    -    PORTD &= ~0b00101111;
    +    DDRB  &= ~0b00111111;
    +    PORTB &= ~0b00111111;
    +    DDRF  &= ~0b00111111;
    +    PORTF &= ~0b00111111;
     }
     
     static void select_row(uint8_t row)
    @@ -145,24 +144,52 @@ static void select_row(uint8_t row)
         // Output low(DDR:1, PORT:0) to select
         switch (row) {
             case 0:
    -            DDRD  |= (1<<0);
    -            PORTD &= ~(1<<0);
    +            DDRB  |= (1<<5);
    +            PORTB &= ~(1<<5);
                 break;
             case 1:
    -            DDRD  |= (1<<1);
    -            PORTD &= ~(1<<1);
    +            DDRB  |= (1<<4);
    +            PORTB &= ~(1<<4);
                 break;
             case 2:
    -            DDRD  |= (1<<2);
    -            PORTD &= ~(1<<2);
    +            DDRB  |= (1<<3);
    +            PORTB &= ~(1<<3);
                 break;
             case 3:
    -            DDRD  |= (1<<3);
    -            PORTD &= ~(1<<3);
    +            DDRB  |= (1<<2);
    +            PORTB &= ~(1<<2);
                 break;
             case 4:
    -            DDRD  |= (1<<5);
    -            PORTD &= ~(1<<5);
    +            DDRB  |= (1<<1);
    +            PORTB &= ~(1<<1);
    +            break;
    +        case 5:
    +            DDRB  |= (1<<0);
    +            PORTB &= ~(1<<0);
    +            break;
    +        case 6:
    +            DDRF  |= (1<<5);
    +            PORTF &= ~(1<<5);
    +            break;
    +        case 7:
    +            DDRF  |= (1<<4);
    +            PORTF &= ~(1<<4);
    +            break;
    +        case 8:
    +            DDRF  |= (1<<3);
    +            PORTF &= ~(1<<3);
    +            break;
    +        case 9:
    +            DDRF  |= (1<<2);
    +            PORTF &= ~(1<<2);
    +            break;
    +        case 10:
    +            DDRF  |= (1<<1);
    +            PORTF &= ~(1<<1);
    +            break;
    +        case 11:
    +            DDRF  |= (1<<0);
    +            PORTF &= ~(1<<0);
                 break;
         }
     }

**Comment by Svalorzen** [comment]
I'm finding through testing that any time I touch the column pins with anything, the debouncing starts (even with row pins unattached). This is both with my finger, or with the cables from the keyboard. To repeat: if I don't attach ANY cables, but a couple on the column side, it still debounces. This is extremely weird...

**Comment by Svalorzen** [comment]
I finally discovered what it was. *My keyboard was a theremin*. So what was happening is that for some reason, the teensy is 100000x more sensible than the chip I had before. For some reason, while there was no direct contact between the cables and the aluminum plate holding the keys, it started to register random presses any time my hand went near the plate (didn't even had to touch it). I've actually made a very cool video where I just move my hand and suddenly the keyboard starts registering debounces.

To fix this I had to connect the ground pin of the teensy to the aluminum plate. Now everything works perfectly. But boy was this a weird problem!!
